contributors/design-proposals/live-and-inplace-vertical-scaling.md

```diff
@@ -0,0 +1,325 @@
+
+2. Services with large amount of state associated with specific instances may encounter temporary loss in access to the specific state information while its corresponding container is being restarted. Or they might encounter non-trivial delays in getting back to expected performance.
+
+3. State may be maintained on node-local storage/devices and restart might place the container/pod on a different node (unless suitably constrained). Might create more fragile scaling option for service and/or additional burden (to constrain placement) on user.
+
+4. Where the restart is not needed, unnecessary load may be created on Kubernetes components that are responsible for pod deletion and creation potentially impacting Kubernetes scalability.
+Resizing running containers of a pod in place could avoid that additional load.
+
+Conversely, resizing without restart may not be the right option for all applications and the right option may also be resource dependent.
+For example, an application might (or have the means to) assess resources available to it only at its initialization time and be unable to adjust its usage if those are dynamically changed.
+Or it might be able to adjust to changes in certain resources (most can for CPU) but not for others (some application have particular difficulty freeing memory they have taken up).
+So we need the means to express and implement the best approach for resizing for each workload and resource.
+
+# Use Cases
+
+* After deploying a service/application through `StatefulSet`, I want to resize, by giving a updated pod spec with new resource sizes, the resources (for now, CPU and memory) allocated to each pod of the statefulset without restarting the pods.
+
+* A (vertical) pod auto-scaler (c.f. [a related issue](https://github.com/kubernetes/features/issues/21)) can utilize this feature, while currently it does resize a pod with restart only.
+
+# Objectives
+
+1. Enable live and in-place resource resizing on a pod.
+2. Add support for live and in-place resource resizing in `StatefulSet` controller.
+
+# API and Usage
+
+To express the policy for resizing in a pod spec, we introduce resource attribute `resizePolicy` with the following choices for value:
+* RestartOnly. (the current behavior, default)
+* LiveResizeable.
+
+This attribute will be available per resource (such as cpu, memory) and so is adequate to indicate whether the workload can handle and prefer a change in each resource’s allocation for it without restarting.
+With potentially multiple containers and multiple resizeable resources for each in a Pod, the response to an update of the pod spec will be determined by the a precedence order among the attribute values with RestartOnly dominating LiveResizeable, i.e., if two resources have been resized in the update to the spec and one of them has a policy of RestartOnly then the pod would be restarted to realize both updates.
```

Comment:
For clarity, can you distinguish Container restart versus pod restart?  From the yaml below, I think you are referring to Container restart since that is all that has resource requirements.  Pod restart terminology to me brings extra baggage for things like init containers.

Reply:
I am also wondering this, a Pod is a collection of containers, if we just adjust resource for container A, shall we restart the Pod or just restart the container A?

Reply:
Actually, our prototype implementation supports container restart (at pod-level resizing). But, we thought that in this proposal it is better to focus on adding a live-and-in-place resizing feature at StatefulSet level, because if we go with container start as well, it seems like to introduce too much things at one time. But, we definitely see the value of Container restart. So, if most of you agree/want to add a Container restart option, we're happy to do that.

Reply:
I could see more value if we could support Container level restart.